### PR TITLE
Feature/lists

### PR DESCRIPTION
**Pull request by SergioEstevao**
Refs #208 
~Gutenberg PR https://github.com/WordPress/gutenberg/pull/14249~
New Gutenberg PR https://github.com/WordPress/gutenberg/pull/14636

simulator screen shot - iphone xr - 2019-03-05 at 22 29 37

This PR implements the list block by only changing components and **not touching** the code of the original web block.

This achieved by improving the implementation of the block-edit context in order to send the isSelected and onFocus props to the RichText components.

For the moment it doesn't implement the indent and outdent buttons, because of a bug in Aztec for iOS that doesn't allow split to work properly when lists are indented.

If testing on iOS please run `yarn preios` to make sure you get the right version of Aztec. 

How to test:
 - Start the demo app 
 - Check the list display properly and you can interact with them and change type
 - Add new list block to the post
 - Check that the list block works properly.

### CODE REVIEW

File: react-native-aztec/src/AztecView.js [hunk 1]
```diff
@@ -146,7 +146,7 @@ class AztecView extends React.Component {
           onContentSizeChange = { this._onContentSizeChange }
           onHTMLContentWithCursor = { this._onHTMLContentWithCursor }
           onSelectionChange = { this._onSelectionChange }
-          onEnter = { this._onEnter }
+          onEnter = { this.onEnter ? this._onEnter : null }
```
Comment (hotchemi) [inline react-native-aztec/src/AztecView.js]: Might be really nitpicking but `undefined` is slightly better?

Comment (Tug) [reply react-native-aztec/src/AztecView.js]: I guess you meant `this.props.onEnter && this._onEnter` here?

Comment (hotchemi) [reply react-native-aztec/src/AztecView.js]: Ah yea!
..and I get used to TS and [only using undefined](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined) but found there's no standard equivalent option for eslint..ignore me😅

**Comment by SergioEstevao** [comment]
Here is update of the current status of this PR:
 
- Insertion of a new list block
- Display of list block already present in the content
- Ability to write inside the lists
- Convert a block list from ordered `ol` to unordered 'ul' using the toolbar
- Indent in is working
- Indent out is not working
- When splitting the block an extra paragraph block is being inserted on the middle of the split.

All the above are working on iOS, but on Android we have some issues:

- Changing from bullets to number list is not working on Android, this is done on JS code so it's strange for me why is not working.
- Aztec android needs to implement the isMultiline prop added to the component. When this boolean parameter is on onEnter prop should only be invoked when an Enter is pressed and the previous line is an empty line.

@koke do you think after the Android issues we should merge this with a feature flag in place?

@hypest do you want to tackle the Android issue?

**Comment by hypest** [comment]
As far as I can see on Android, adding a new item to the list by tapping Enter at the end of an item is broken. Instead of adding a new bullet, the block is actually split. Haven't debugged this yet but, @SergioEstevao are you referring to that too when you say `Aztec android needs to implement the isMultiline prop added to the component. When this boolean parameter is on onEnter prop should only be invoked when an Enter is pressed and the previous line is an empty line.`?

> @hypest do you want to tackle the Android issue?

Yeap, added it to my plans for this sprint 👍

**Comment by SergioEstevao** [comment]
Yes it's exactly that! When the multiline property is true pressing Enter
shouldn't send an onEnter event up that causes the split.

**Comment by koke** [comment]
> When splitting the block an extra paragraph block is being inserted on the middle of the split.

Can you explain this one?

I think it’s fine to merge with a feature flag when Android is working. If indent out is not working, I’d disable both buttons (in/out) until that works

**Comment by SergioEstevao** [comment]
@koke here is a video:

ListSplits

**Comment by koke** [comment]
I'm seeing the same on the web (except for the stray bullet in the second list):

lists-web

**Comment by SergioEstevao** [comment]
> I'm seeing the same on the web (except for the stray bullet in the second list):
> 
> lists-web

Should I open a bug on the GB repo ? :)

**Comment by koke** [comment]
No, I think that's the right behaviour. I just checked and Google Docs, Apple Pages, and Microsoft Word do the same thing.

**Comment by SergioEstevao** [comment]
@Tug do you mind to test this. If testing on iOS please run `yarn preios` to make sure you get the right version of Aztec.

File: src/block-management/block-holder.js [hunk 2]
```diff
@@ -238,6 +239,7 @@ export default compose( [
 			isLastBlock,
 			isSelected,
 			name,
+			clientId,
```
Comment (Tug) [inline src/block-management/block-holder.js]: `withSelect` only add new props to the component, so it should not be necessary to explicitly add `clientId` which is already a prop given from the parent component here.

File: src/block-management/block-holder.js [hunk 3]
```diff
@@ -181,6 +181,7 @@ export class BlockHolder extends React.Component<PropsType, StateType> {
 				insertBlocksAfter={ this.insertBlocksAfter }
 				mergeBlocks={ this.mergeBlocks }
 				onCaretVerticalPositionChange={ this.props.onCaretVerticalPositionChange }
+				clientId={ this.props.clientId }
```
Comment (Tug) [inline src/block-management/block-holder.js]: That's good, it should have been added earlier 👍

File: react-native-aztec/src/AztecView.js [hunk 4]
```diff
@@ -157,10 +157,13 @@ class AztecView extends React.Component {
...
           onSelectionChange = { this._onSelectionChange }
-          onEnter = { this._onEnter }
+          onEnter = { this.props.onEnter && this._onEnter }
+          // IMPORTANT: the onFocus events are thrown away as these are handled by onPress() in the upper level.
+          // It's necessary to do this otherwise onFocus may be set by `{...otherProps}` and thus the onPress + onFocus
+          // combination generate an infinite loop as described in https://github.com/wordpress-mobile/gutenberg-mobile/issues/302
           onFocus = { this._onAztecFocus } 
           onBlur = { this._onBlur }
-          onBackspace = { this._onBackspace }
+		  onBackspace = { this._onBackspace }
```
Comment (Tug) [inline react-native-aztec/src/AztecView.js]: Minor but we're mixing tabs and spaces here, I think we should convert the whole file to tabs for more consistency

**Pull request merged**